src/adapter.rs

```diff
@@ -522,4 +768,197 @@ impl BluetoothAdapter for Empty{
+       self.0.get_address()
+    }
+
+    fn set_address(&self, address: String) -> Result<(), Box<Error>> {
+        match self {
+            &BluetoothAdapter::Mock(ref fake_adapter) => fake_adapter.set_address(address),
+            _ => Err(Box::from(NOT_SUPPORTED_ON_REAL_ERROR)),
+        }
+    }
+
+    fn get_name(&self) -> Result<String, Box<Error>> {
+       self.0.get_name()
+    }
+
+    fn set_name(&self, name: String) -> Result<(), Box<Error>> {
+       self.0.set_name()
+    }
+
+    fn get_alias(&self) -> Result<String, Box<Error>> {
+       self.0.get_alias()
+    }
+
+    fn set_alias(&self, alias: String) -> Result<(), Box<Error>> {
+        self.0.set_alias()
+    }
+
+    fn get_class(&self) -> Result<u32, Box<Error>> {
+        self.0.get_class()
+    }
+
+    fn set_class(&self, class: u32) -> Result<(), Box<Error>> {
+        self.0.set_class()
+    }
+
+    fn is_powered(&self) -> Result<bool, Box<Error>> {
+       self.0.is_powered()
+    }
+
+    fn set_powered(&self, powered: bool) -> Result<(), Box<Error>> {
+        self.0.set_powered()
+    }
+
+    fn is_present(&self) -> Result<bool, Box<Error>> {
+        self.0.is_present()
+    }
+
+    fn set_present(&self, present: bool) -> Result<(), Box<Error>> {
+        self.0.set_present()
+    }
+
+    fn is_discoverable(&self) -> Result<bool, Box<Error>> {
+        self.0.is_discoverable()
+    }
+
+    
+    fn set_discoverable(&self, discoverable: bool) -> Result<(), Box<Error>> {
+       self.0.set_discoverable()
+    }
+
+    fn is_pairable(&self) -> Result<bool, Box<Error>> {
+        self.0.is_pairable()
+    }
+
+    fn set_pairable(&self, pairable: bool) -> Result<(), Box<Error>> {
+        self.0.set_pairable()
+    }
+
+    fn get_pairable_timeout(&self) -> Result<u32, Box<Error>> {
+        self.0.get_pairable_timeout()
+    }
+
+    fn set_pairable_timeout(&self, timeout: u32) -> Result<(), Box<Error>> {
+        self.0.set_pairable_timeout()
+    }
+
+    fn get_discoverable_timeout(&self) -> Result<u32, Box<Error>> {
+        self.0.get_discoverable_timeout()
+    }
+
+    fn set_discoverable_timeout(&self, timeout: u32) -> Result<(), Box<Error>> {
+        self.0.set_discoverable_timeout()
+    }
+
+    fn is_discovering(&self) -> Result<bool, Box<Error>> {
+        self.0.is_discovering()
+    }
+
+    fn set_discovering(&self, discovering: bool) -> Result<(), Box<Error>> {
+        self.0.set_discovering()
+    }
+
+    fn set_can_start_discovery(&self, can_start_discovery: bool) -> Result<(), Box<Error>> {
+        self.0.set_can_start_discovery()
+    }
+
+    fn set_can_stop_discovery(&self, can_stop_discovery: bool) -> Result<(), Box<Error>> {
+        self.0.set_can_stop_discovery()
+    }
+
+    fn create_discovery_session(&self) -> Result<BluetoothDiscoverySession, Box<Error>> {
+        //BluetoothDiscoverySession::create_session(self.clone())
+        let test_session = try!(FakeBluetoothDiscoverySession::create_session(fake_adapter));
```

Comment:
self.0 instead of fake_adapter?

Reply:
when I use self.0 and run cargo build --features bluetooth-test, it throws an error, cannot move out of borrowed content

Reply:
And I cannot use Copy trait to resolve this as Arc<FakaebluetoothAdapter> does not use Copy trait

Reply:
`self.0.clone()` should work fine.